Pattern matches are non-exhaustive

Compiling git `HEAD` I get:
```
src/PyF/Formatters.hs:177:9: warning: [-Wincomplete-uni-patterns]
    Pattern match(es) are non-exhaustive
    In a pattern binding:
        Patterns not matched:
            FractionalRepr Negative [] [] []
            FractionalRepr Negative [] [] (_:_)
            FractionalRepr Negative [] (_:_) []
            FractionalRepr Negative [] (_:_) (_:_)
            ...
    |
177 |     let (FractionalRepr Positive aa bb cc) = reprFractional fmt precision (abs f)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/PyF/Formatters.hs:191:22: warning: [-Wincomplete-uni-patterns]
    Pattern match(es) are non-exhaustive
    In a pattern binding:
        Patterns not matched:
            ([], [], (_:_))
            ([], (_:_), (_:_))
            ((_:_), [], (_:_))
            ((_:_), (_:_), (_:_))
    |
191 |       Percent -> let (a, b, "") = splitFractional (Numeric.showFFloatAlt precision (iAbs * 100) "") in (a, b, "%")
```
Even if these extra options are basically impossible, the compiler can't prove that so the dev has to satisfy the compiler.

--- a/PyF.cabal
+++ b/PyF.cabal
@@ -35,7 +35,9 @@ library
                      , parsec >= 3.1.13 && < 3.2
                      , mtl >= 2.2.2 && < 2.4
                      , ghc >= 8.6.1 && < 9.7
-                     , ghc-boot >= 8.6.1 && < 9.7
+  if impl(ghc < 9.2.1)
+    build-depends:
+                       ghc-boot >= 8.6.1 && < 9.7
   hs-source-dirs: src
   ghc-options: -Wall -Wunused-packages -Wincomplete-uni-patterns
   default-language:    Haskell2010

--- a/src/PyF/Formatters.hs
+++ b/src/PyF/Formatters.hs
@@ -173,9 +173,9 @@ reprFractional :: (RealFloat f) => Format t t' 'Fractional -> Maybe Int -> f -> 
 reprFractional fmt precision f
   | isInfinite f = Infinite sign (upperIt "inf")
   | isNaN f = NaN (upperIt "nan")
-  | isNegativeZero f =
-    let (FractionalRepr Positive aa bb cc) = reprFractional fmt precision (abs f)
-     in FractionalRepr Negative aa bb cc
+  | isNegativeZero f = case reprFractional fmt precision (abs f) of
+       FractionalRepr Positive aa bb cc -> FractionalRepr Negative aa bb cc
+       other -> error $ "reprFractional (isNegativeZero f): The impossible happened : " ++ show other ++ ". Please open an issue at https://github.com/guibou/PyF/issues/"
   | otherwise = FractionalRepr sign decimalPart fractionalPart suffixPart
   where
     upperIt s = case fmt of
@@ -188,7 +188,9 @@ reprFractional fmt precision f
       Fixed -> splitFractional (Numeric.showFFloatAlt precision iAbs "")
       Exponent -> overrideExponent precision $ splitFractionalExp (Numeric.showEFloat precision iAbs "")
       Generic -> splitFractionalExp (Numeric.showGFloatAlt precision iAbs "")
-      Percent -> let (a, b, "") = splitFractional (Numeric.showFFloatAlt precision (iAbs * 100) "") in (a, b, "%")
+      Percent -> case splitFractional (Numeric.showFFloatAlt precision (iAbs * 100) "") of
+                  (a, b, "") -> (a, b, "%")
+                  other -> error $ "reprFractional (format): The impossible happened : " ++ show other ++ ". Please open an issue at https://github.com/guibou/PyF/issues/"
       Alternate fmt' -> format fmt'
       Upper fmt' ->
         let (a, b, c) = format fmt'

--- a/src/PyF/Internal/Meta.hs
+++ b/src/PyF/Internal/Meta.hs
@@ -35,7 +35,12 @@ import HsLit
 import qualified Data.ByteString as B
 import qualified Language.Haskell.TH.Syntax as GhcTH
 import qualified Language.Haskell.TH.Syntax as TH
+
+#if MIN_VERSION_ghc(9,6,0)
+import PyF.Internal.ParserEx (fakeSettings)
+#else
 import PyF.Internal.ParserEx (fakeLlvmConfig, fakeSettings)
+#endif
 
 #if MIN_VERSION_ghc(9,6,0)
 import GHC.Types.SourceText (il_value, rationalFromFractionalLit,SourceText(..))
@@ -49,7 +54,11 @@ import GHC.Data.FastString
 #if MIN_VERSION_ghc(9,2,0)
 import GHC.Utils.Outputable (ppr)
 import GHC.Types.Basic (Boxity(..))
+#if MIN_VERSION_ghc(9,6,0)
+import GHC.Types.SourceText (FractionalLit)
+#else
 import GHC.Types.SourceText (il_value, rationalFromFractionalLit, FractionalLit)
+#endif
 import GHC.Driver.Ppr (showSDoc)
 #else
 import GHC.Utils.Outputable (ppr, showSDoc)
@@ -190,7 +199,7 @@ toExp d (Expr.ExplicitTuple _ (map unLoc -> args) boxity) = ctor tupArgs
 #else
     tupArgs = case traverse toTupArg args of
       Nothing -> error "Tuple section are not supported by template haskell < 8.10"
-      Just args -> fmap (toExp d) args
+      Just args' -> fmap (toExp d) args'
 #endif
 
 {- ORMOLU_ENABLE -}
@@ -247,10 +256,10 @@ toExp _ (HsProjection _ fields) = TH.ProjectionE (fmap (unpackFS . unLoc . hflLa
 toExp dynFlags e = todo "toExp" (showSDoc dynFlags . ppr $ e)
 
 todo :: (HasCallStack, Show e) => String -> e -> a
-todo fun thing = error . concat $ [moduleName, ".", fun, ": not implemented: ", show thing]
+todo fun thing = error . concat $ [moduleName, ".", fun, ": not implemented: ", show thing, "Please open an issue at https://github.com/guibou/PyF/issues"]
 
 noTH :: (HasCallStack, Show e) => String -> e -> a
-noTH fun thing = error . concat $ [moduleName, ".", fun, ": no TemplateHaskell for: ", show thing]
+noTH fun thing = error . concat $ [moduleName, ".", fun, ": no TemplateHaskell for: ", show thing, "Please open an issue at https://github.com/guibou/PyF/issues"]
 
 moduleName :: String
 moduleName = "PyF.Internal.Meta"

--- a/src/PyF/Internal/Meta.hs
+++ b/src/PyF/Internal/Meta.hs
@@ -37,6 +37,10 @@ import qualified Language.Haskell.TH.Syntax as GhcTH
 import qualified Language.Haskell.TH.Syntax as TH
 import PyF.Internal.ParserEx (fakeLlvmConfig, fakeSettings)
 
+#if MIN_VERSION_ghc(9,6,0)
+import GHC.Types.SourceText (il_value, rationalFromFractionalLit,SourceText(..))
+#endif
+
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Types.SrcLoc
 import GHC.Types.Name
@@ -45,7 +49,7 @@ import GHC.Data.FastString
 #if MIN_VERSION_ghc(9,2,0)
 import GHC.Utils.Outputable (ppr)
 import GHC.Types.Basic (Boxity(..))
-import GHC.Types.SourceText (il_value, rationalFromFractionalLit,SourceText(..))
+import GHC.Types.SourceText (il_value, rationalFromFractionalLit, FractionalLit)
 import GHC.Driver.Ppr (showSDoc)
 #else
 import GHC.Utils.Outputable (ppr, showSDoc)
@@ -68,6 +72,7 @@ import GHC.Stack
 
 #if MIN_VERSION_ghc(9,2,0)
 -- TODO: why this disapears in GHC >= 9.2?
+fl_value :: FractionalLit -> Rational
 fl_value = rationalFromFractionalLit
 #endif
 
@@ -250,13 +255,13 @@ noTH fun thing = error . concat $ [moduleName, ".", fun, ": no TemplateHaskell f
 moduleName :: String
 moduleName = "PyF.Internal.Meta"
 
+baseDynFlags :: [GhcTH.Extension] -> DynFlags
+baseDynFlags exts = foldl xopt_set dynFlags enable
+  where
+    enable = GhcTH.TemplateHaskellQuotes : exts
 #if MIN_VERSION_ghc(9,6,0)
-dynFlags = defaultDynFlags fakeSettings
+    dynFlags = defaultDynFlags fakeSettings
 #else
-dynFlags = defaultDynFlags fakeSettings fakeLlvmConfig
+    dynFlags = defaultDynFlags fakeSettings fakeLlvmConfig
 #endif
 
-baseDynFlags :: [GhcTH.Extension] -> DynFlags
-baseDynFlags exts =
-  let enable = GhcTH.TemplateHaskellQuotes : exts
-   in foldl xopt_set dynFlags enable

--- a/src/PyF/Internal/Parser.hs
+++ b/src/PyF/Internal/Parser.hs
@@ -24,14 +24,13 @@ import GHC.Utils.Outputable
 import GHC.Utils.Error
 #endif
 
-#if MIN_VERSION_ghc(9,2,0)
+#if MIN_VERSION_ghc(9,4,0)
+#elif MIN_VERSION_ghc(9,2,0)
 import qualified GHC.Parser.Errors.Ppr as ParserErrorPpr
 #endif
 
 #if MIN_VERSION_ghc(9,0,0)
 import qualified GHC.Types.SrcLoc as SrcLoc
-#else
-import qualified SrcLoc
 #endif
 
 #if MIN_VERSION_ghc(9,0,0)
@@ -70,7 +69,6 @@ parseExpression initLoc s dynFlags =
 #elif MIN_VERSION_ghc(8,10,0)
     PFailed PState{loc=srcLoc, messages=msgs} ->
 #else
-    -- TODO: check for pattern failure
     PFailed _ (SrcLoc.srcSpanEnd -> SrcLoc.RealSrcLoc srcLoc) doc ->
 #endif
 
@@ -83,9 +81,9 @@ parseExpression initLoc s dynFlags =
                     $ map errMsgDiagnostic
                     $ sortMsgBag Nothing
                     $ getMessages $ errorMessages
-                line = SrcLoc.srcLocLine srcLoc
+                line' = SrcLoc.srcLocLine srcLoc
                 col = SrcLoc.srcLocCol srcLoc
-            in Left (line, col, err)
+            in Left (line', col, err)
 #elif MIN_VERSION_ghc(9,3,0)
             let
                 err = renderWithContext defaultSDocContext
@@ -118,4 +116,10 @@ parseExpression initLoc s dynFlags =
                 line = SrcLoc.srcLocLine srcLoc
                 col = SrcLoc.srcLocCol srcLoc
             in Left (line, col, err)
+#endif
+
+#if MIN_VERSION_ghc(8,10,0)
+#elif MIN_VERSION_ghc(8,6,0)
+    -- Only here to satisfy GHC checker which was not able to consider this as total with GHC <8.10
+    PFailed _ _ _ -> error "The impossible happen: this case is not possible"
 #endif

--- a/src/PyF/Internal/ParserEx.hs
+++ b/src/PyF/Internal/ParserEx.hs
@@ -212,5 +212,6 @@ parseExpression initLoc s flags =
     POk s e -> unP (runECP_P e) s
     PFailed ps -> PFailed ps
 #else
+parseExpression :: RealSrcLoc -> String -> DynFlags -> ParseResult (LHsExpr GhcPs)
 parseExpression initLoc = parse initLoc Parser.parseExpression
 #endif

--- a/src/PyF/Internal/PythonSyntax.hs
+++ b/src/PyF/Internal/PythonSyntax.hs
@@ -36,8 +36,12 @@ import PyF.Internal.Meta
 import qualified PyF.Internal.Parser as ParseExp
 import Text.Parsec
 import Data.Data (Data)
+
+#if MIN_VERSION_ghc(9,6,0)
+-- For some reasons, theses function are not exported anymore by some others
 import Data.Functor (void)
 import Control.Monad (replicateM_)
+#endif
 
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Types.SrcLoc
@@ -152,7 +156,7 @@ parseExpressionString = do
   -- Special case for "::", we want to parse it as part of an expression,
   -- unless it may be the end of the format field (':'), followed by a padding
   -- char (':') followed by a padding specifier.
-  res <- some ((try (string "::" <* notFollowedBy (oneOf "<>=^"))) <|> (pure <$> noneOf (charClosing : ":" :: String)))
+  res <- some (try (string "::" <* notFollowedBy (oneOf "<>=^")) <|> (pure <$> noneOf (charClosing : ":" :: String)))
   pure $ concat res
 
 replacementField :: Parser Item

--- a/src/PyF/Internal/QQ.hs
+++ b/src/PyF/Internal/QQ.hs
@@ -31,10 +31,11 @@ import Control.Monad.Reader
 import Data.Data (Data (gmapQ), Typeable, cast)
 import Data.Kind
 import Data.List (intercalate)
-import Data.Maybe (catMaybes, fromMaybe)
+import Data.Maybe (catMaybes, fromMaybe, isJust)
 import Data.Proxy
 import Data.String (fromString)
-import GHC (GenLocated (L), moduleNameString)
+
+import GHC (moduleNameString)
 
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Tc.Utils.Monad (addErrAt)
@@ -100,14 +101,11 @@ import Text.Parsec
 import Text.Parsec.Error
   ( errorMessages,
     messageString,
-    newErrorMessage,
-    setErrorPos,
     showErrorMessages,
   )
-import Text.Parsec.Pos (newPos, initialPos)
+import Text.Parsec.Pos (initialPos)
 import Text.ParserCombinators.Parsec.Error (Message (..))
 import Unsafe.Coerce (unsafeCoerce)
-import Data.Maybe (isJust)
 
 -- | Configuration for the quasiquoter
 data Config = Config
@@ -192,7 +190,7 @@ findFreeVariables item = allNames
     f :: forall a. (Data a, Typeable a) => a -> [Located RdrName]
     f e = case cast @_ @(HsExpr GhcPs) e of
 #if MIN_VERSION_ghc(9,2,0)
-      Just (HsVar _ l@(L a b)) -> [L (locA a) (unLoc l)]
+      Just (HsVar _ l@(L a _)) -> [L (locA a) (unLoc l)]
 #else
       Just (HsVar _ l) -> [l]
 #endif
@@ -230,9 +228,9 @@ lookupName n = case n of
 doesExists :: (b, RdrName) -> Q (Maybe (String, b))
 doesExists (loc, name) = do
   res <- lookupName name
-  case res of
-    False -> pure (Just ("Variable not in scope: " <> show (toName name), loc))
-    True -> pure Nothing
+  if res
+    then pure Nothing
+    else pure (Just ("Variable not in scope: " <> show (toName name), loc))
 
 -- | Check that all variables used in 'Item' exists, otherwise, fail.
 checkVariables :: [Item] -> Q (Maybe (SrcSpan, String))

--- a/src/PyF/Internal/QQ.hs
+++ b/src/PyF/Internal/QQ.hs
@@ -35,7 +35,6 @@ import Data.Maybe (catMaybes, fromMaybe, isJust)
 import Data.Proxy
 import Data.String (fromString)
 
-import GHC (moduleNameString)
 
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Tc.Utils.Monad (addErrAt)
@@ -44,16 +43,24 @@ import GHC.Types.Name (occNameString)
 #else
 import OccName
 import TcRnTypes (TcM)
-import TcSplice (lookupThName_maybe)
 import TcRnMonad (addErrAt)
+#endif
+
+#if MIN_VERSION_ghc(9,6,0)
+#else
+import GHC (moduleNameString)
 #endif
 
 #if MIN_VERSION_ghc(9,3,0)
 import GHC.Tc.Errors.Types
 import GHC.Types.Error
+import GHC.Utils.Outputable (text)
+
+#if MIN_VERSION_ghc(9,6,0)
+#else
 import GHC.Driver.Errors.Types
 import GHC.Parser.Errors.Types
-import GHC.Utils.Outputable (text)
+#endif
 #endif
 
 
@@ -61,7 +68,6 @@ import GHC.Utils.Outputable (text)
 #if MIN_VERSION_ghc(9,0,0)
 import GHC.Types.Name.Reader
 #else
-import FastString
 import RdrName
 #endif
 
@@ -73,7 +79,6 @@ import GHC.Hs.Pat as Pat
 import HsExpr as Expr
 import HsExtension as Ext
 import HsPat as Pat
-import HsLit
 #endif
 
 #if MIN_VERSION_ghc(9,0,0)
@@ -82,10 +87,8 @@ import GHC.Types.SrcLoc
 import SrcLoc
 #endif
 
-#if MIN_VERSION_ghc(8,10,0)
+#if MIN_VERSION_ghc(9,2,0)
 import GHC.Hs
-#else
-import HsSyn
 #endif
 
 import GHC.TypeLits

--- a/test/Spec.hs
+++ b/test/Spec.hs
@@ -81,6 +81,7 @@ instance (a ~ a') => IsLabel a (V a') where
 showV :: KnownSymbol a => V a -> String
 showV = show
 
+globalName :: String
 globalName = "Valérian"
 
 spec :: Spec

